spec.md

```diff
@@ -0,0 +1,70 @@
+# spec
+
+The DHT needs to provide the following mappings:
+* `NodeId` `=>` `addr`, `port`
+* `writers` `=>` `[peer_id]`
+* `cid` `=>` `[peer_id]`
+
+## NodeId
+
+* `node_id` is like `libp2p/core/peer_id` but it strives to offer first-class support for anti-sybil mechanisms to slow id generation. I eventually want the proof of work method to be configurable based on some defined shared state (like the blockchain pow algorithm), but this is not my first priority.
```

Comment:
that would require a consensus mechansim, which is out of scope for the dht. you could provide a function to set the difficulty from a higher layer.

Reply:
It's not out of scope for s-kademlia, which theoretically requires a provably difficult and temper-proof mechanism to limit NodeId generation in a given time period. If we're just hashing public keys, then I can have a few machines doing that all the time to generate a set of public keys for which the respective hash maintains the requisite number of trailing zeros. Then, I can just sybil attack the network with the **pre-generated set of public keys** whenever I want. 

The current design is dumb -- it's not robust against sybil attacks and I don't want to pretend it is. Let's assume that it is provably difficult via benchmarking or hardware testing. This just means that normal nodes won't be able to generate a new `NodeId` in a reasonable amount of time and, if the puzzle is hard enough, it will restrict the network to the nodes that have the processing power to generate sets of keys and whoever they want to share their pre-generated keys with.

Reply:
I could organize the code such that a new `NodeHandler` generates a `NodeId` with a `difficulty` input set by some `Protocol` in the `impl NodeHandler` `new` method. This makes a lot of sense and I'm thinking about it.

Reply:
As I said you can provide an API to set the difficulty. But then don't forget that if you change the difficulty dynamically, everyone needs to generate a new identity

Reply:
why? I can change the difficulty and still accept old identities. I just need to come to consensus on the set of identities (as discussed two comments above).

Reply:
Chicken and egg. How do you plan to reach consensus without a dht?

Reply:
Can we please not mix consensus, bitswap and dht? They're completely different things and should be treated as such.

Reply:
You don't need to use the same DHT for consensus on the set of nodes.

**You said please so I'll drop this for now**